### ISSUE TITLE
Get rid of IDs for synchronization primitives, make pthread shims not leaky

### ISSUE THREAD

**Issue by RalfJung** [rust-lang/miri#3967]
We currently store the state of our synchronization primitives with a layer of indirection: there are lists in `SynchronizationObjects` and then we use IDs to index into these lists. These lists keep growing while the program runs and never get cleaned up, which isn't very clean and also makes @saethlin unhappy. ;)

As of https://github.com/rust-lang/miri/pull/3966, we no longer store these IDs in machine memory, so there isn't actually a good reason to still use IDs. Instead, the `PthreadMutex` used by `pthread_mutex_t` can just directly store a `sync::Mutex`. That way, when the allocation that contains the mutex is freed, we also free all the extra data associated with the mutex, thus avoiding memory leaks.

The main trouble here is functions like `mutex_unlock(&mut self, id: MutexId)`: these would want to take a reference to the `Mutex` instead, but since they also take a mutable reference to the entire machine state, that can't work! We could carefully arrange things to avoid overlapping references here... but probably the easier approach is to use reference counting. So we'd make `PthreadMutex` store a `MutexRef` (which is a newtype for `Rc<RefCell<Mutex>>`) and the function signatures would look like
```rust
fn mutex_unlock(&mut self, id: &MutexRef) -> ...
```
https://github.com/rust-lang/miri/pull/3971 does this for `Futex` so that can serve as a model.

Currently we are also using these IDs in `BlockReason`, but that's just a sanity check, so it's okay to remove them there.

Also as part of this, we could finally fix the leaks in `pthread_mutex_destroy` and friends.

**Comment by tiif** [comment]
I still have a few things to clear in my todo list, but I'd like to work on a different part of the codebase, so, 

@rustbot claim

**Comment by RalfJung** [comment]
Note that this is blocked on https://github.com/rust-lang/miri/pull/3966

**Comment by tiif** [comment]
@RalfJung Is this also blocked on #3971? If not I could start working on this soon :)

**Comment by RalfJung** [comment]
> Is this also blocked on https://github.com/rust-lang/miri/pull/3971?

No. That PR can just be helpful as a guide for what this could look like -- there we have `Futex` and `FutexRef`, so here we'd have e.g. `Mutex` and `MutexRef`.

**Comment by tiif** [comment]
@rustbot release-assignment

I have quite a few things on my queue right now. Current ``SynchronizationObjects`` is left with three lists, which I expect all of them could be removed. 

```rust
pub struct SynchronizationObjects {
    rwlocks: IndexVec<RwLockId, RwLock>,
    condvars: IndexVec<CondvarId, Condvar>,
    pub(super) init_onces: IndexVec<InitOnceId, InitOnce>,
}

```

**Comment by RalfJung** [comment]
https://github.com/rust-lang/miri/pull/4002 can serve as a good template for what removing one of these lists looks like.
